### PR TITLE
Some corrections

### PR DESCRIPTION
**Pull request by bogem**
Simplify, delete redundant code, add docs, add more verbosity to error messages.
More in commits.

Updates https://github.com/spf13/cobra/issues/27

### CODE REVIEW

File: flag.go [hunk 1]
```diff
@@ -406,18 +406,29 @@ func (f *FlagSet) Set(name, value string) error {
...
 	err := flag.Value.Set(value)
 	if err != nil {
-		return err
+		var flagName string
+		if len(flag.Shorthand) > 0 && len(flag.ShorthandDeprecated) == 0 {
+			flagName = fmt.Sprintf("-%s, --%s", flag.Shorthand, flag.Name)
+		} else {
+			flagName = fmt.Sprintf("--%s", flag.Name)
+		}
+		return fmt.Errorf("invalid argument %q for %q flag: %v", value, flagName, err)
```
Comment (eparis) [inline flag.go]: I'm trying to decide if `invalid argument` is always correct.  Maybe just `unable to set argument`

File: flag.go [hunk 2]
```diff
@@ -406,18 +406,29 @@ func (f *FlagSet) Set(name, value string) error {
...
 		f.actual = make(map[NormalizedName]*Flag)
 	}
 	f.actual[normalName] = flag
 	f.orderedActual = append(f.orderedActual, flag)
-	flag.Changed = true
+	f.sortedActual = f.sortedActual[:0]
+
 	if len(flag.Deprecated) > 0 {
-		fmt.Fprintf(os.Stderr, "Flag --%s has been deprecated, %s\n", flag.Name, flag.Deprecated)
+		fmt.Fprintf(f.out(), "Flag --%s has been deprecated, %s\n", flag.Name, flag.Deprecated)
```
Comment (eparis) [inline flag.go]: you sure about this change? This does feel like stderr to me...

File: flag.go [hunk 3]
```diff
@@ -904,53 +889,71 @@ func (f *FlagSet) parseLongArg(s string, args []string, fn parseFunc) (a []strin
...
 		value = args[0]
 		outArgs = args[1:]
 	} else {
+		// '-f' (arg was required)
 		err = f.failf("flag needs an argument: %q in -%s", c, shorthands)
 		return
 	}
-	err = fn(flag, value, shorthands)
+
+	if len(flag.ShorthandDeprecated) > 0 && containsShorthand(shorthands, flag.Shorthand) {
```
Comment (eparis) [inline flag.go]: I don't think I understand this. If we get here and `len(flag.ShorthandDeprecated) > 0` we know this was called via the deprecated shorthand. No? Why do we need this `containsShorthand` test?

File: flag.go [hunk 4]
```diff
@@ -904,53 +889,71 @@ func (f *FlagSet) parseLongArg(s string, args []string, fn parseFunc) (a []strin
...
+		fmt.Fprintf(f.out(), "Flag shorthand -%s has been deprecated, %s\n", flag.Shorthand, flag.ShorthandDeprecated)
```
Comment (eparis) [inline flag.go]: I think this should be stderr.

File: flag.go (hunk 3 above)
Comment (bogem) [inline flag.go]: I just copied it from [setFlag](https://github.com/spf13/pflag/blob/master/flag.go#L858) function.

Actually I asked this question for myself too. I thought, there can be special cases, where it be useful. If you delete containsShorthand, the tests are passed.

File: flag.go (hunk 4 above)
Comment (bogem) [inline flag.go]: I can't understand, when I should use f.out() or stderr. Can you please explain me?

File: flag.go (hunk 1 above)
Comment (eparis) [inline flag.go]: Fair enough. Up to you if you want to make it better.

File: flag.go (hunk 2 above)
Comment (eparis) [inline flag.go]: ok, i'll buy it. All of our stdout/stderr handling need some major rework...

File: flag.go (hunk 3 above)
Comment (eparis) [inline flag.go]: I think by putting it here we don't need it anymore. Lets kill it.

File: flag.go (hunk 1 above)
Comment (bogem) [inline flag.go]: I think "invalid argument" is clear and good enough for this error.

**Comment by bogem** [comment]
@eparis looks good to you?

**Pull request merged**